File: haskell/p20.hs
-- Remove the K'th element from a list.

-- Using "splitAt"
f1 :: Int -> [a] -> ([a], a)
f1 n xs = let (a, x:b) = splitAt (pred n) xs in (a ++ b, x)

-- By index (inefficient)
f2 :: Int -> [a] -> ([a], a)
f2 n xs = (take (pred n) xs ++ drop n xs, xs !! (pred n))

-- Recursion
f3 :: Int -> [a] -> ([a], a)
f3 n (x:xs)
 | n > 1 = let (a, b) = f3 (pred n) xs in (x : a, b)
 | otherwise = (xs, x)

-- Tail recursion
f4 :: Int -> [a] -> ([a], a)
f4 = f []
    where f acc n (x:xs)
            | n > 1 = f (x : acc) (pred n) xs
            | otherwise = (reverse acc ++ xs, x)

File: clojure/P20.clj
; P20 Remove the K'th element from a list.

; Using "split-at"
(defn f1 [n l]
  (let [[a [x & b]] (split-at (dec n) l)]
    [(concat a b) x]))

; By index (inefficient)
(defn f2 [n l]
  [(concat (take (dec n) l) (drop n l)) (nth l (dec n))])

; Recursion
(defn f3 [n l]
  (let [[h & t] l]
    (if (> n 1)
      (let [[a b] (f3 (dec n) t)]
        [(cons h a) b])
      [t h])))

; Tail recursion
(defn f4 [n l]
  (loop [acc '() i n [h & t] l]
    (if (> i 1)
      (recur (cons h acc) (dec i) t)
      [(concat (reverse acc) t) h])))